progress.go

```diff
@@ -0,0 +1,165 @@
+package scylla_cdc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gocql/gocql"
+)
+
+type ProgressReporter interface {
+	// MarkProgress saves information about the last successfully processed
+	// cdc log record. If the reader is restarted, it should resume reading
+	// from this shard starting after the last marked position.
+	MarkProgress(progress Progress) error
+}
+
+type ProgressManager interface {
+	// GetCurrentGenerationTime returns the time of the generation that was
+	// saved as being currently processed. If there was no such information
+	// saved, it returns a zero value for Time.
+	GetCurrentGeneration() (time.Time, error)
+
+	// StartGeneration is called when a generation is discovered and the reader
+	// should start processing the next generation.
+	StartGeneration(gen time.Time) error
+
+	// EndGeneration marks all streams in the generation as "done".
+	EndGeneration(gen time.Time) error
+
+	// GetProgress retrieves information about the progress of given shard,
+	// in a given table.
+	GetProgress(gen time.Time, table string, streamID StreamID) (Progress, error)
+
+	// SaveProgress stores information about the last cdc log record which was
+	// processed successfully. If the reader is restarted, it should resume
+	// work for this shard from the last saved
+	SaveProgress(gen time.Time, table string, streamID StreamID, progress Progress) error
+}
+
+type Progress struct {
+	Time gocql.UUID
+}
+
+// NoProgressManager does not persist the progress at all. It can be used
+// when saving progress is not necessary.
+type NoProgressManager struct{}
+
+func (*NoProgressManager) GetCurrentGeneration() (time.Time, error) {
+	return time.Time{}, nil
+}
+
+func (*NoProgressManager) StartGeneration(gen time.Time) error {
+	return nil
+}
+
+func (*NoProgressManager) EndGeneration(gen time.Time) error {
+	return nil
+}
+
+func (*NoProgressManager) GetProgress(gen time.Time, table string, streamID StreamID) (Progress, error) {
+	return Progress{}, nil
+}
+
+func (*NoProgressManager) SaveProgress(gen time.Time, table string, streamID StreamID, progress Progress) error {
+	return nil
+}
+
+type TableBackedProgressManager struct {
+	session           *gocql.Session
+	progressTableName string
+
+	// TTL to use when writing progress for a shard (a week by default).
+	// TODO: maybe not? maybe we should clean up this data manually?
+	// Progress data may be large if generations are very large
+	ttl int32
+}
+
+func NewTableBackedProgressManager(session *gocql.Session, progressTableName string) (*TableBackedProgressManager, error) {
+	tbpm := &TableBackedProgressManager{
+		session:           session,
+		progressTableName: progressTableName,
+
+		ttl: 7 * 24 * 60 * 60, // 1 week
```

Comment:
shouldn't  this be configurable? (pass as parameter to `New...`)
value `0` should be possible which does not set ttl